aea/mail/base.py

```diff
@@ -482,104 +478,67 @@ def connection_status(self) -> ConnectionStatus:
+                await asyncio.wait_for(self._stop(), timeout=60)
                 self._connection_status.is_connected = False
             except (CancelledError, Exception):
+                logger.exception("Exception on disconnect:")
                 raise AEAConnectionError("Failed to disconnect the multiplexer.")
 
-    def _run_loop(self):
+    async def _stop(self) -> None:
         """
-        Run the asyncio loop.
+        Stop the multiplexer.
 
-        This method is supposed to be run only in the Multiplexer thread.
+        Stops recv and send loops.
+        Disconnect every connection.
         """
-        logger.debug("Starting threaded asyncio loop...")
-        asyncio.set_event_loop(self._loop)
-        self._out_queue = asyncio.Queue()
-        self._loop.run_forever()
-        logger.debug("Asyncio loop has been stopped.")
-
-    def _start_loop_threaded_if_not_running(self):
-        """Start the multiplexer."""
-        if not self._loop.is_running() and not self._thread.is_alive():
-            self._thread.start()
-        logger.debug("Multiplexer started.")
-
-    def _stop(self):
-        """Stop the multiplexer."""
-        if self._recv_loop_task is not None and not self._recv_loop_task.done():
-            self._recv_loop_task.cancel()
+        logger.debug("Stopping multiplexer...")
+        await cancel_and_wait(self._recv_loop_task)
 
         if self._send_loop_task is not None and not self._send_loop_task.done():
             # send a 'stop' token (a None value) to wake up the coroutine waiting for outgoing envelopes.
-            asyncio.run_coroutine_threadsafe(
-                self.out_queue.put(None), self._loop
-            ).result()
-            self._send_loop_task.cancel()
-
-        if self._connect_all_task is not None:
-            self._connect_all_task.cancel()
-        if self._disconnect_all_task is not None:
-            self._disconnect_all_task.cancel()
+            await self.out_queue.put(None)
+            await cancel_and_wait(self._send_loop_task)
 
         for connection in [
             c
             for c in self.connections
             if c.connection_status.is_connected or c.connection_status.is_connecting
         ]:
-            asyncio.run_coroutine_threadsafe(
-                connection.disconnect(), self._loop
-            ).result()
-
-        if self._loop.is_running() and not self._thread.is_alive():
-            self._loop.call_soon_threadsafe(self._loop.stop)
-            self._loop.stop()
-        elif self._loop.is_running() and self._thread.is_alive():
-            self._loop.call_soon_threadsafe(self._loop.stop)
-            self._thread.join()
+            await connection.disconnect()
         logger.debug("Multiplexer stopped.")
 
-    async def _connect_all(self):
+    async def _connect_all(self) -> None:
         """Set all the connection up."""
         logger.debug("Start multiplexer connections.")
         connected = []  # type: List[PublicId]
```

Comment:
notice: below we are connecting each connection _serially_, and not concurrently.

Do you see any pros/cons in making it concurrently connecting? One thing is that it's harder to debug in case some connection does not connect.

Reply:
i think it's another task. current refactoring made only for better asyncio interface, not for internal optimizations

but yes, it can be done in parallel, but have to concern about better exception info provided.